test/functional/p2p_invalid_messages.py

```diff
@@ -248,6 +250,34 @@ def test_oversized_headers_msg(self):
         size = MAX_HEADERS_RESULTS + 1
         self.test_oversized_msg(msg_headers([CBlockHeader()] * size), size)
 
+    def test_invalid_pow_headers_msg(self):
+        self.log.info("Test headers message with invalid proof-of-work is logged as misbehaving and disconnects peer")
+        blockheader_tip_hash = self.nodes[0].getbestblockhash()
+        blockheader_tip = from_hex(CBlockHeader(), self.nodes[0].getblockheader(blockheader_tip_hash, False))
+
+        # send valid headers message first
+        assert_equal(self.nodes[0].getblockchaininfo()['headers'], 0)
+        blockheader = CBlockHeader()
+        blockheader.hashPrevBlock = int(blockheader_tip_hash, 16)
+        blockheader.nTime = int(time.time())
+        blockheader.nBits = blockheader_tip.nBits
+        blockheader.rehash()
+        while not blockheader.hash.startswith('0'):
+            blockheader.nNonce += 1
+            blockheader.rehash()
+        good_peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        good_peer.send_and_ping(msg_headers([blockheader]))
+        assert_equal(self.nodes[0].getblockchaininfo()['headers'], 1)
```

Comment:
nit: Could also check the connected header hash
```python
assert any(x["hash"] == blockheader.hash and x["status"] == "headers-only" for x in self.nodes[0].getchaintips())
```

Reply:
Good idea, done. Decided to access the `getchaintips()` result at index 0 (works since the result is always returned in decreasing order of height) in order to take use of `assert_equal`, so if this ever fails, we see the expected/actual values immediately.